wren-ui/src/apollo/server/managers/dataSourceSchemaDetector.ts

```diff
@@ -0,0 +1,288 @@
+  resolveSchemaChange(type: string): Promise<void>;
+}
+
+export default class DataSourceSchemaDetector
+  implements IDataSourceSchemaDetector
+{
+  public ctx: IContext;
+  public projectId: number;
+
+  constructor({ ctx, projectId }: { ctx: IContext; projectId: number }) {
+    this.ctx = ctx;
+    this.projectId = projectId;
+  }
+
+  public async detectSchemaChange() {
+    const diffSchema = await this.getDiffSchema();
+    if (diffSchema) {
+      this.addSchemaChange(diffSchema);
+    }
+  }
+
+  public async resolveSchemaChange(type: string) {
+    const schemaChangeType = camelCase(type) as SchemaChangeType;
+    const supportedTypes = [
+      SchemaChangeType.DELETED_TABLES,
+      SchemaChangeType.DELETED_COLUMNS,
+    ];
+    if (supportedTypes.includes(schemaChangeType)) {
+      const lastSchemaChange =
+        await this.ctx.schemaChangeRepository.findLastSchemaChange(
+          this.projectId,
+        );
+      const changes = lastSchemaChange?.change[schemaChangeType];
+      const isResolved = lastSchemaChange?.resolve[schemaChangeType];
+
+      if (isResolved !== false) {
+        throw new Error(
+          `Schema change "${schemaChangeType}" has nothing to resolve.`,
+        );
+      }
+
+      // Handle resolve deleted tables
+      if (schemaChangeType === SchemaChangeType.DELETED_TABLES) {
+        const affectedTableNames = changes.map((table) => table.name);
+        logger.debug(
+          `Start to remove tables "${affectedTableNames}" from models.`,
+        );
+        await this.ctx.modelRepository.deleteAllBySourceTableNames(
+          affectedTableNames,
+        );
+        await updateResolveToSchemaChange(this.ctx);
+      }
+
+      // Handle resolve deleted table columns
+      if (schemaChangeType === SchemaChangeType.DELETED_COLUMNS) {
+        const models = await this.ctx.modelRepository.findAllBy({
+          projectId: this.projectId,
+        });
+        const affectedModels = models.filter(
+          (model) =>
+            changes.findIndex(
+              (table) => table.name === model.sourceTableName,
+            ) !== -1,
+        );
+        await Promise.all(
+          affectedModels.map(async (model) => {
+            const affectedColumnNames = changes
+              .find((table) => table.name === model.sourceTableName)
+              .columns.map((column) => column.name);
+            logger.debug(
+              `Start to remove columns "${affectedColumnNames}" from model "${model.referenceName}".`,
+            );
+            return await this.ctx.modelColumnRepository.deleteAllBySourceColumnNames(
+              model.id,
+              affectedColumnNames,
+            );
+          }),
+        );
+        await updateResolveToSchemaChange(this.ctx);
+      }
+
+      async function updateResolveToSchemaChange(ctx: IContext) {
+        await ctx.schemaChangeRepository.updateOne(lastSchemaChange.id, {
+          ...lastSchemaChange,
```

Comment:
Why do you need to update other columns? I think only "resolve" column needs to update